tests/subsys/net/lib/nrf_cloud/cloud/src/main.c

```diff
@@ -0,0 +1,1880 @@
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	zassert_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the initialized state after initialization");
+	(void)start_connection_poll();
+
+	int ret = start_connection_poll();
+
+	zassert_equal(-EINPROGRESS, ret, "return should be -EINPROGRESS when already active");
+}
+
+/* Verify that the function will succeed
+ * if the cloud is successfully initialized
+ */
+ZTEST(nrf_cloud_test, test_start_connection_poll_success)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	int ret = start_connection_poll();
+
+	zassert_ok(ret, "return should be 0 when success");
+}
+
+/* TEST NRF_CLOUD_RUN */
+
+/* Verify that the function will change the cloud disconnect
+ * event to not success if the connection to the cloud failed
+ */
+ZTEST(nrf_cloud_test, test_cloud_run_cloud_connecting_failed)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud failed */
+	nct_connect_fake.custom_fake = fake_nct_connect__invalid;
+
+	(void)start_connection_poll();
+
+	/* Wait for the connecting event for 2s, timeout after that */
+	wait_counter_t = 10;
+	while (--wait_counter_t && !atomic_get(&transport_connecting_t)) {
+		k_sleep(K_MSEC(200));
+	}
+	zassert_not_equal(0, wait_counter_t, "cloud connecting timed out");
+
+	zassert_not_equal(NRF_CLOUD_CONNECT_RES_SUCCESS, status_t,
+		"status should not be success when transport failed");
+}
+
+/* Verify that the function will change the cloud
+ * disconnect event to closed by remote if timeout
+ * has expired and no connection to the cloud
+ */
+ZTEST(nrf_cloud_test, test_cloud_run_timed_out_connection_closed_by_remote)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud successfully */
+	nct_connect_fake.custom_fake = fake_nct_connect__succeeds;
+	nct_disconnect_fake.custom_fake = fake_nct_disconnect__succeeds;
+	/* These fakes are for the poll thread */
+	nct_socket_get_fake.custom_fake = fake_nct_socket_get_data;
+	nct_keepalive_time_left_fake.custom_fake = fake_nct_keepalive_time_left_get;
+
+	/* Transport process needs to be failed */
+	nct_process_fake.custom_fake = fake_nct_process__fails;
+	/* Change poll_fake to indicate timeout expired */
+	poll_fake.custom_fake = fake_poll__zero;
+
+	(void)start_connection_poll();
+
+	/* Wait for the events */
+	wait_for_connecting_and_disconnected_poll_thread();
+
+	zassert_equal(NRF_CLOUD_DISCONNECT_CLOSED_BY_REMOTE, status_t,
+		"status should be closed by remote");
+}
```

Comment:
This definitely feels like something that should test on mocked MQTT, rather than just directly set all the error codes in nct/fsm to what we would expect in the event of a timeout